main/src/ca/uwaterloo/flix/language/phase/OccurrenceAnalyzer.scala

```diff
@@ -0,0 +1,396 @@
+/*
+ * Copyright 2022 Magnus Madsen, Anna Krogh, Patrick Lundvig, Christian Bonde
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package ca.uwaterloo.flix.language.phase
+
+import ca.uwaterloo.flix.api.Flix
+import ca.uwaterloo.flix.language.CompilationMessage
+import ca.uwaterloo.flix.language.ast.LiftedAst.Expression
+import ca.uwaterloo.flix.language.ast.OccurrenceAst.Occur
+import ca.uwaterloo.flix.language.ast.OccurrenceAst.Occur._
+import ca.uwaterloo.flix.language.ast.{LiftedAst, OccurrenceAst, Symbol}
+import ca.uwaterloo.flix.util.Validation
+import ca.uwaterloo.flix.util.Validation.ToSuccess
+
+/**
+ * The occurrence analyzer collects information on variable usage
+ */
+object OccurrenceAnalyzer {
```

Comment:
We can call it OccurrenceAnalyzer for now, maybe later we can find a better name.